File: problem-1-FloodFill.py
"""
Problem:
https://leetcode.com/problems/flood-fill/

Approach: BFS Solution

1. change the cell corresponding to sr,sc and add it to the queue
2. while queue is not empty , take out the elemnts from queue, change the color of all the neighbours to newColor and add those neighbours to the que

"""

# Time Complexity : O(m*n) where m is the no of rows and n is no of columns
# Space Complexity : O(m*n)
# Did this code successfully run on Leetcode : yes	
# Any problem you faced while coding this : No


from collections import deque
class Solution:
    newColor = None
    oldColor = None
     
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if image==None or len(image)==0 or image[sr][sc]==newColor:
            return image
        
        self.oldColor = image[sr][sc]
        self.newColor = newColor
        que = deque()
        
        #change color and add to the queue
        image[sr][sc] =  self.newColor
        que.append((sr,sc))
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        
        while(que):
            cur = que.popleft()
            for direction in directions:
                r = cur[0]+direction[0]
                c=  cur[1]+direction[1]
                if(r >=0 and r<=(len(image)-1) and c>=0 and c<=(len(image[0])-1) and image[r][c]==self.oldColor):
                    image[r][c]= self.newColor
                    que.append((r,c))
        
        return image


"""
Approach: DFS Solution
"""

# Time Complexity : O(m*n) where m is the no of rows and n is no of columns
# Space Complexity : O(m*n) where m is the no of rows and n is no of columns
# Did this code successfully run on Leetcode : yes	
# Any problem you faced while coding this : No

class Solution:
    newColor = None
    oldColor = None
    
    def dfs(self,image,sr,sc):
        #base case
        if(sr<0 or sr>(len(image)-1) or sc<0 or sc> (len(image[0])-1) or image[sr][sc]!=self.oldColor):
            return
        
        
        #logic
        image[sr][sc]=self.newColor
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        for direction in directions:
            r = sr+direction[0]
            c=  sc+direction[1]
            self.dfs(image,r,c)
        
    
    
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if image==None or len(image)==0 or image[sr][sc]==newColor:
            return image
        
        self.oldColor = image[sr][sc]
        self.newColor = newColor
        
        self.dfs(image,sr,sc)
        return image

File: Problem1_FloodFill.java
// Time Complexity : o(mn)
// Space Complexity :o(mn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    
    // using BFS
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        
          if(image==null||image.length==0||image[sr][sc]==newColor)
              return image;
        
        int m=image.length;
        int n=image[0].length;
        int oldColor=image[sr][sc];
        image[sr][sc]=newColor;
        
        Queue<int[]> queue=new LinkedList<>();
        queue.add(new int[]{sr,sc});
        int[][] dirs={{0,1},{0,-1},{-1,0},{1,0}};
        
        while(!queue.isEmpty())
        {
            int[] curr=queue.poll();
            
            for(int[] dir:dirs)
            {
                  int nR=dir[0]+curr[0];
                int nC=dir[1]+curr[1];
                if(nR>=0 && nR<m && nC>=0 && nC<n && image[nR][nC]==oldColor)
                    {
                          image[nR][nC]=newColor;
                            queue.add(new int[]{nR,nC});
                }
            }
            
        }
        return image;
    }
    
    // USING DFS Recursion
      int m;
      int n;
      int oldColor;
      int[][] dirs;
      
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        
          if(image==null||image.length==0||image[sr][sc]==newColor)
              return image;
        
         m=image.length;
         n=image[0].length;
         oldColor=image[sr][sc];
         dirs={{0,1},{0,-1},{-1,0},{1,0}};
        dfs(image,sr,sc,newColor);
        return image;
        
    }
    
    private void dfs(int[][] image,int sr,int sc,int newColor)
    {
         if(nR>=0 && nR<m && nC>=0 && nC<n && image[nR][nC]==oldColor)
         {
             return;
         }
        image[sr][sc]=newColor;
        
        for(int[] dir:dirs)
        {
            int r=sr+dir[0];
            int c=sc+dir[1];
            
            dfs(image,r,c,newColor);
        }
        
    }
}
